Attempt to fix block synchronizer request storm by tweaking the latching.

alsrdn opened the pull request:
Attempt to fix block synchronizer request storm by tweaking the latching.

Don't decrement the latch when responses for old requests that are no longer needed (are received in a state that expects other data).
For example if the synchonizer is the `HaveWeakFinality` state, it will try to fetch the block body from some peers. The latch will be incremented by `num_peers`. If a late finality signature response comes in this state it will decrement the latch. If multiple late responses come, the latch can become 0, and the synchronizer would try to fetch the block body again even if no response for the initial requests were received yet.

Added a test for this also. It's a bit ugly and copy-pasty but I think it showcases the issue.

Attempted fix for: https://github.com/casper-network/casper-node/issues/4202

alsrdn committed:
block_synchronizer: fix latching decrement

Don't decrement the latch when responses for old requests that are no
longer needed are received.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
block_synchronizer/tests: add new test scenario for checking latching

Add test scenario to check block synchonizer latching when old responses
are received in a state that no longer needs them.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

EdHastingsCasperAssociation commented on node/src/components/block_synchronizer/block_builder.rs:
```diff
@@ -456,6 +461,23 @@ impl BlockBuilder {
 
+    pub(super) fn waiting_for_block_header(&self) -> bool {
+        match &self.acquisition_state {
+            BlockAcquisitionState::Initialized(..) => true,
```
an if let for these methods would probably be less annoying.

alsrdn replied:
I've given some thought and while I agree it's annoying using a wildcard will be less helpful if we refactor this logic in the future to add/remove states. Having a match here would help guide future changes (at least force us to reason about it) so I would prefer if we leave it as is. WDYT?

alsrdn committed:
block_synchronizer: fix latching decrement for deploy and exec results

Don't decrement latch for late deploy fetch responses. Also don't
decrement latch for late execution results.

Add tests for checking latching scenarios with late deploys and exec
results.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

rafal-ch commented on node/src/components/block_synchronizer/block_builder/tests.rs:
```diff
@@ -25,38 +25,38 @@ fn handle_acceptance() {
     // Builder acceptance for needed signature from ourselves.
     assert!(builder
-        .handle_acceptance(None, Ok(Some(Acceptance::NeededIt)))
+        .handle_acceptance(None, Ok(Some(Acceptance::NeededIt)), true)
```
Can you add a single case to this test when the `should_unlatch` is `false`?

alsrdn replied:
Added a new test in https://github.com/casper-network/casper-node/pull/4205/commits/896653b1abdade591bd96cc092faa4199e98fbd0 and renamed the old one for more granular testing.

rafal-ch reviewed (approved):
LGTM :+1: 
No blockers

alsrdn committed:
tests/block_builder: add test for unlatching builder on acceptance

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn commented:
bors r+
